automation/publish_signals.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2022 Searis AS
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package automation
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"runtime/debug"
+
+	"github.com/clarify/clarify-go"
+	"github.com/clarify/clarify-go/query"
+	"github.com/clarify/clarify-go/views"
+	"golang.org/x/exp/maps"
+)
+
+const (
+	selectSignalsPageSize  = 1000
+	publishSignalsPageSize = 500
+)
+
+// Annotations used by automation tasks.
+const (
+	AnnotationPrefix                    = "clarify/clarify-go/"
+	AnnotationPublisherName             = AnnotationPrefix + "publisher/name"
+	AnnotationPublisherTransformVersion = AnnotationPrefix + "publisher/transform-version"
+	AnnotationPublisherSignalID         = AnnotationPrefix + "publisher/signal-id"
+	AnnotationPublisherSignalAttributes = AnnotationPrefix + "publisher/signal-attributes"
+)
+
+// LogOptions describe the options for operation logs.
+type LogOptions struct {
+	// Verbose, when true, enables detailed logs, such as full JSON summaries of
+	// operations.
+	Verbose bool
+
+	// Out describe the destination writer for logs. If unset, all logging is
```

Comment:
maybe write logs by default to stdout?

Reply:
Probably you mean Stderr. I did so initially. but it makes it harder to disable the logs, as you can't disable it by setting it to nil. Also, it's common in go to let logging be disabled by default in library packages.

E.g. if you have a program using structured logging, you would not want unstructured logs to end up among the output without you having actively enabled it.

Reply:
It's currently set by default in the automation template _binary_:

https://github.com/clarify/template-clarify-automation/blob/b134f06441797f0628c241d484691ef2a6418c28/commands.go#L102

Reply:
Actually I meant `stdout`, but now I understand that it's best practice to log to `stderr` 👍